mlfinlab/labeling/matrix_flags.py

```diff
@@ -0,0 +1,148 @@
+# pylint: disable=no-self-use
+# pylint: disable=unnecessary-comprehension
+"""
+Matrix Flag labeling method.
+"""
+
+import pandas as pd
+import numpy as np
+
+
+class MatrixFlagLabels:
+    """
+    The Matrix Flag labeling method is featured in the paper: Cervelló-Royo, R., Guijarro, F. and Michniuk, K., 2015.
+    Stock market trading rule based on pattern recognition and technical analysis: Forecasting the DJIA index with
+    intraday data.
+
+    The method of applying a matrix template was first introduced, and explained in greater detail, in the paper:
+    Leigh, W., Modani, N., Purvis, R. and Roberts, T., 2002. Stock market trading rule discovery using technical
+    charting heuristics.
+
+    Cervelló-Royo et al. expand on Leigh et al.'s work by proposing a new bull flag pattern which ameliorates some
+    weaknesses in Leigh's original template. Additionally, he applies this bull flag labeling method to intraday
+    candlestick data, rather than just closing prices.
+
+    To find the total weight for a given day, the current price as well as the preceding window days number of prices is
+    used. The data window is split into 10 buckets each containing a chronological tenth of the data window. Each point
+    in 1 bucket is put into a decile corresponding to a position in a column based on percentile relative to the entire
+    data window. Bottom 10% on lowest row, next 10% on second lowest row etc.
+    The proportion of points in each decile is reported to finalize the column. The first tenth of the data is
+    transformed to the leftmost column, the next tenth to the next column on the right and so on until finally a 10 by
+    10 matrix is achieved. This matrix is then multiplied element-wise with the 10 by 10 template, and the sum of all
+    columns is the total weight for the day. If desired, the user can specify a threshold to determine positive and
+    negative classes. The value of the threshold depends on how strict of a classifier the user desires, and the
+    allowable values based on the template matrix.
+    """
+
+    def __init__(self, data, window):
+        """
+        :param data: (pd.Series) Price data for one stock.
+        :param window: (int) Length of preceding data window used when generating the fit matrix for one day.
+        """
```

Comment:
I'm also thinking, maybe we should allow whole dataframes as inputs to label whole datasets. What do you think @ruifanp?

Reply:
Maybe. I considered it but I kept it at a series to keep the code a bit simpler, and unlike the other methods I don't really think having a dataframe as allowable input would add much real utility to the user, since I assumed that since this is a pretty specific labeling method the user should know what stock they want to use it for. What do you think, @Jackal08?

Reply:
I'm also not sure if it makes sense to use the same threshold (if categorical labels) for different stocks, since the choice of what threshold to use is subjective and depends on the situation,

Reply:
@ruifanp True that the code is simpler this way.

The threshold can be passed as a series and be specific for each asset, but it's probably an additional feature to be added in the future if we decide it's useful.